Add gitea2john support

isaac-app-dev opened the pull request:
This script allows conversion of Gitea PBKDF2-SHA256 password hashes into John-compatible format.

Creation of this module was driven by a Hack The Box Retired Machine called [Titanic](https://app.hackthebox.com/machines/Titanic). Probably rare in the wild and more plausible in a CTF style environment, but was run into and figured it may be good to add if it works for John.

**Expected input (e.g. from SQLite dump):**

username|hashed salt|hashed password|password algorithm

**Example Output:**

username:$pbkdf2-sha256$50000$base64-salt$base64-hash-TRIMMED

where `50000` is the iteration count listed in the password algorithm

**Usage Example:**

```zsh
sqlite3 gitea.db 'SELECT name, salt, passwd, passwd_hash_algo FROM user;' | gitea2john.py > john.hashes
```

```zsh
 john --format=pbkdf2-hmac-sha256 john.hashes --wordlist=rockyou.txt
```

The module handles hex decoding and then re-encoding to base64 to satisfy John. In addition, to satisfy John's 32 Byte Hash limit restriction needs, this module performs a trimming of oversized hashes (e.g. dkLen=50), and outputs:
    username$pbkdf2-sha256$iteration_count$salt_b64$hash-b64-trimmed

NOTE: Default iteration count for Gitea PBKDF2 algorithm is 50000. However users can change this configuration to their desired needs. This module also dynamically handles extracting the iterations count when building the hash format.

Tested with Python 3.13.3

isaac-app-dev committed:
add gitea2john.py: Convert Gitea PBKDF2 Hashes for John

solardiz commented:
Thank you!

Can you please share a full working input line for (re)testing this script? Maybe include it in the comment inside the script.

Can you please squash the two commits here into one, and force-push? (only if this is easy for you to do - that is, if you use `git` on the command-line)

It would also be good to add an entry to `doc/NEWS`. Can be part of that one commit above or separate.

solardiz commented:
I also wonder whether we could/should support this via prepare() in the format, instead of adding a conversion script like this. But we can make this change separately from this PR. The script approach has the maybe-advantage as it also serves as documentation of availability of this functionality. Otherwise we need to add a file under `doc`.

isaac-app-dev commented:
> Thank you!
> 
> Can you please share a full working input line for (re)testing this script? Maybe include it in the comment inside the script.
> 
> Can you please squash the two commits here into one, and force-push? (only if this is easy for you to do - that is, if you use `git` on the command-line)
> 
> It would also be good to add an entry to `doc/NEWS`. Can be part of that one commit above or separate.

will make a second commit for doc/NEWS  and will include a working input line to test it. I have squashed the other two for now, let me know if you want me to squash them all into one, I can do that for sure nbd

isaac-app-dev commented:
> Thank you!
> 
> Can you please share a full working input line for (re)testing this script? Maybe include it in the comment inside the script.
> 
> Can you please squash the two commits here into one, and force-push? (only if this is easy for you to do - that is, if you use `git` on the command-line)
> 
> It would also be good to add an entry to `doc/NEWS`. Can be part of that one commit above or separate.

Before pushing the next commit, wanted to confirm -> for your request to add an entry in doc/NEWS, should I rather create a README.gitea to go along with the other entries there? doc/NEWS is kinda more John centric

solardiz commented:
Please also fix this, as reported by our CI whitespace-errors job:

```
run/gitea2john.py:66: trailing whitespace.
+    
```

> for your request to add an entry in doc/NEWS, should I rather create a README.gitea to go along with the other entries there? doc/NEWS is kinda more John centric

It'd be good if you also create a `README-Gitea` file, but a `NEWS` entry is desired anyway. We do have `NEWS` entries about added tools.

At a later time, if we go with the `prepare()` approach, maybe your script will be removed, but `README-Gitea` will stay - edited to reflect the usage change.

isaac-app-dev commented:
> Please also fix this, as reported by our CI whitespace-errors job:
> 
> ```
> run/gitea2john.py:66: trailing whitespace.
> +    
> ```
> 
>  > for your request to add an entry in doc/NEWS, should I rather create a README.gitea to go along with the other entries there? doc/NEWS is kinda more John centric
> 
> It'd be good if you also create a `README-Gitea` file, but a `NEWS` entry is desired anyway. We do have `NEWS` entries about added tools.
> 
> At a later time, if we go with the `prepare()` approach, maybe your script will be removed, but `README-Gitea` will stay - edited to reflect the usage change.

used some sed commands to try to pull all trailing white spaces out. I am using vscode on Mac, so not sure if something is adding a trail on upload but will see how tests go.

I did add a new example input line in the helper doc of the script for users to test with. Let me know if you have any questions on it. 

I only added the NEWS entry for now per the original request but happy to add the README of the team desires it prior merge.

solardiz commented:
The whitespace error is still there, and now you certainly need to squash the commits (after you fix for real). While we work on a PR, we shouldn't be mentioning any fixes of issues within this same PR in commit messages - when you do that, it just tells you that those fixes should have been rolled into the original commits instead.

Also, looks like there's no newline character after the very last line of the script.

And the `NEWS` entry should be at the end of the list for the upcoming release, right after other entries for 2025. And we wrap lines in that file at 79.

solardiz commented:
> used some sed commands to try to pull all trailing white spaces out.

Apparently, it's on line 66 (empty line before the line loop) - maybe you can "see" it in the editor and deleted it when you go to that line number?

isaac-app-dev commented:
> > used some sed commands to try to pull all trailing white spaces out.
> 
> Apparently, it's on line 66 (empty line before the line loop) - maybe you can "see" it in the editor and deleted it when you go to that line number?

tried, but vscode has given me this headache before, will resolve it, I will also squash all commits including resolving the news entry also

solardiz commented:
> I did add a new example input line in the helper doc of the script for users to test with. Let me know if you have any questions on it.

I don't see it. Maybe you forgot to add the edited file to the commit?

isaac-app-dev commented:
> > I did add a new example input line in the helper doc of the script for users to test with. Let me know if you have any questions on it.
> 
> I don't see it. Maybe you forgot to add the edited file to the commit?

Squashed all the commits to one, sorry for my slowness, been a minute since contributing. So resolved items

1. Found the whitespace issue, had a couple spaces lingering on a new line and removed them. Checked through the file and should be good
2. the sample command with example input is in the helper doc section at the top of the python file. Will be visible with the -h option as well
3. Added the new line at the end of the python file for the linter (rookie mistake on my end)
4. Resolved the NEWS entry. I moved it to the bottom in the desired format.

solardiz commented:
Thank you! There are still a few things I'd adjust, but I'm going to merge this as-is now, then adjust.

isaac-app-dev commented:
> Thank you! There are still a few things I'd adjust, but I'm going to merge this as-is now, then adjust.

💯, sounds good, feel free to let me know of anything else, happy to update and do another pr if needed

solardiz commented:
I wonder if this entire script could be replaced with a more elaborate SQLite query in a documentation file. SQLite is rather powerful, see e.g. https://sqlite.org/draft/base64.html

isaac-app-dev commented:
> I wonder if this entire script could be replaced with a more elaborate SQLite query in a documentation file. SQLite is rather powerful, see e.g. https://sqlite.org/draft/base64.html

I thought about this but the database is not always sqlite3. Gitea supports Postgres, MySQL and mariadb as well. Just depends on the config. To be fair, more likely sqlite3 will be present for small installations. That being said if we do decide to go that route we definitely could. We just have to include extracting for the other dbs as well.
